doc/metadata-server/README.md

```diff
@@ -0,0 +1,327 @@
+
+### Draft of the metadata schema
+
+Following [CIP-0026 convention](https://github.com/cardano-foundation/CIPs/tree/master/CIP-0026):
+
+- We use [_Blake2b-256_](https://github.com/cardano-foundation/CIPs/tree/master/CIP-0026#hash-function),
+
+- We use [this encoding](https://github.com/cardano-foundation/CIPs/tree/master/CIP-0026#attestation-signatures) to construct messages for signing.
+
+#### Instance metadata
+
+The `id` of the metadata can be derived from the transaction:
+
+```js
+contract = serialize(Contract)
+
+roleDistribution = [{ "name": <string>, "pubKeyHash": <string> }]
+
+insId = hash(
+  hash(CBOR(contract)) +
+  hash(CBOR(roleDistribution)) +
+  validatorHash
+)
+```
+
+`serialize(Contract)` - `JSON` serialized version of the Marlowe expression from the `Tx`
+`validatorHash` - a marlowe validator hash bytestring
+`CBOR` - [function](https://github.com/cardano-foundation/CIPs/tree/master/CIP-0026#attestation-signatures) which encodes its `JSON` input into binary
+
+We can define now the suggested shape of the `instanceMetadata`:
+
+```js
+insMetadata = {
+  metadata = {
+    shortDescription: <string>,
+    longDescription: <string>,
+    choiceDescriptions: [{ id: <string>, description: <string> }],
+    roleDescriptions: [{ id: <string>, description: <string> }],
+    <other properties>
+  },
+  "signatures": <string | array>
+}
+```
+
+All properties of `metadata` value are optional but we require that this map is not empty. `signatures` are optional too but if they are present we consider them as [adequate](# verification-of-contract-metadata) to the transaction if they sign data together with its `insId`:
+
+```js
+signature = sign(hash(insId + CBOR(metadata)))
+```
+
+#### Template metadata
+
+```js
+extContractId = hash(CBOR(serialize(extContract)))
+
+extMetadata = {
+  template = {
+    contract: <extContract>,
+    metadata: {
+      shortDescription: <string>,
+      longDescription: <string>,
+      choiceDescriptions: [{ id: <string>, description: <string> }],
+      roleDescriptions: [{ id: <string>, description: <string> }],
+      valueDescriptions: [{ id: <string>, description: <string> }],
+
+      <other properties>
+    }
+  },
+  "params": [{ id: <string>, description: <string> }]
+  "signatures": <string | array>
+}
+```
+`signatures` and `params` are optional as well as `template.contract`. `template.metadata` should not be empty. We consider signature to be relevant to the transaction when:
+
+```js
+signature = sign(hash(extContractId + CBOR(extMetadata.template.metadata)))
+```
+
+and
+
+```js
+txContract == fillTemplate(extMetadata.template.contract, extMetadata.params)
+```
+
+#### Tx metadata schema
+
+```js
+contractMetadata = {"instance": <insMetadata>, "extended": <extMetadata>}
+
+txMetadata = {
+  "1564": {
+      "on": <contractMetadata>
+      "off": {
+        url: <string>
```

Comment:
Maybe add "hash" here? However, NFT metadata usually just uses the "ipfs://" prefix along with the hash.